edge-modules/azure-monitor/src/ModuleClientWrapper.cs

```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Client;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Devices.Edge.Azure.Monitor
+{
+
+    public class ModuleClientWrapper
+    {
+        public ModuleClient Inner;
+        ITransportSettings[] transportSettings;
+
+        public ModuleClientWrapper(ModuleClient moduleClient, ITransportSettings[] transportSettings)
+        {
+            this.Inner = moduleClient;
+            this.transportSettings = transportSettings;
+
+            moduleClient.ProductInfo = Constants.ProductInfo;
+        }
+
+        public static async Task<ModuleClientWrapper> BuildModuleClientWrapperAsync(ITransportSettings[] transportSettings)
+        {
+            ModuleClient moduleClient = await ModuleClient.CreateFromEnvironmentAsync(transportSettings);
+            moduleClient.ProductInfo = Constants.ProductInfo;
+            await moduleClient.OpenAsync();
+
+            return new ModuleClientWrapper(moduleClient, transportSettings);
+        }
+
+        public async Task RecreateClient(CancellationToken ct)
+        {
+            try
+            {
+                this.Inner.Dispose();
+                this.Inner = await ModuleClient.CreateFromEnvironmentAsync(this.transportSettings);
+                this.Inner.ProductInfo = Constants.ProductInfo;
+                await this.Inner.OpenAsync();
+
+                LoggerUtil.Writer.LogInformation("Closed and re-established connection to IoT Hub");
+            }
+            catch (Exception e)
+            {
+                LoggerUtil.Writer.LogWarning($"Failed closing and re-establishing connection to IoT Hub: {e.ToString()}");
```

Comment:
If something in the try block throws, then `this.Inner` will be in a weird state (at least, it won't be open). Should you clean it up here (maybe by calling `Dispose` on it)?

Reply:
It should get disposed the next time the client is recreated. I can call it in the catch, but is it possible that this `Dispose()` then fails too and the exception bubbles up and causes more problems?